### DISCUSSION TITLE
Allow `spawn(<machine>)` to accept an optional starting state

### DISCUSSION THREAD

**Discussion by Alex-Yakubovsky** [statelyai/xstate#3946] [General] [answered]
## What
It'd be handy to set the initial state when calling `spawn(<machine>)`, something like

```ts
spawn(myMachine, <SpawnOptions>, { state: <desired initial state> })
```

## Problem I'm Trying to Solve
I'm building out a small game in which there is a `gameMachine` and a `playerMachine` (each player gets an instance of `playerMachine`). The `gameMachine` spawns a `playerMachine` actor whenever a player joins. The game can be in the `lobby` state or in the `playing` state, if a player joins while the game hasn't started yet, their initial state is `"waiting"`. I'd like for players to also be able to join while the game is being playing, in which case the initial state for the player that join mid game machine should be `"playing"`. Also open to suggestions if I'm approaching this wrong 🙂

**Comment by Andarist** [comment]
Where that "desired initial state" is coming from? How do you get to know it within your app?

**Reply by Alex-Yakubovsky** [reply]
In my case, the desired initial state of the player would be based off of the state of the game.

Here's a minimal example that hopefully illustrates my goal:

```ts
const playerMachine = createMachine({
  states: {
    waiting: { ... },
    playing: { ... }
  }
})

const gameMachine = createMachine({
  context: {
    playerRefs: []      
  }.
  states: {
    lobby: {
      on: { 
        PLAYER_JOINED: { actions: "addWaitingPlayer" } 
      },
    },
    // Assume this is a board game, we are in the tabletop state when the game has begun
    tabletop: {
      // Allows new players to join mid game
      on: { PLAYER_JOINED: { actions: "addPlayingPlayer" } }
    }
  }
},
{
  actions: {
    addWaitingPlayer: assign({ playerRefs: (ctx) => [...ctx.playerRefs, spawn(playerMachine, undefined, { state: 'waiting' })] }),
    addPlayingPlayer: assign({ playerRefs: (ctx) => [...ctx.playerRefs, spawn(playerMachine, undefined, { state: 'playing' })] })
  }
})
```

**Reply by Alex-Yakubovsky** [reply]
^ this would be running server side

**Reply by Andarist** [reply]
Wouldn't a conditional initial state in the `playerMachine` make for a better model here? Smth like:
```ts
initial: [{
  cond: (ctx) => ctx.isPlaying,
  target: 'playing'
}, {
  target: 'waiting'
}]
```

**Reply by Alex-Yakubovsky** [reply]
That would work, my only thought here is that `ctx.isPlaying` and the the `"playing"` state represent the same information, it'd be nice to have a single source of truth. 

A computed initial state could be handy:

```ts
initial: (ctx, init) => init.isPlaying ? 'playing' : 'waiting'
```

or with a similar api as `.withConfig` and `.withContext`; `.withInitial`. So I could do something like:

```ts
const playingPlayerMachine = playerMachine.withInitial('playing').withContext({ .... })
spawn(playingPlayerMachine)
```

**Reply by Andarist** [reply, marked as answer]
The `ctx.isPlaying` was just an example of smth that u could base your decision on. Your `init` proposal~ is pretty much implemented in v5 as `input` (we dont - yet - have dynamic~ initial states there though but there is a plan to introduce them)

**Reply by Alex-Yakubovsky** [reply]
gotcha, I'll keep an eye out for that!
